Mu2eKinKal/src/LoopHelixFit_module.cc

```diff
@@ -347,88 +488,103 @@ namespace mu2e {
-          if(chcolptr != &chcol)
-            throw cet::exception("RECO")<<"mu2e::KKHelixFit: inconsistent ComboHitCollection" << std::endl;
-          // next, build straw hits and materials from these
-          KKSTRAWHITCOL strawhits;
-          strawhits.reserve(strawHitIdxs.size());
-          KKSTRAWXINGCOL strawxings;
-          strawxings.reserve(strawHitIdxs.size());
-          if(kkfit_.makeStrawHits(*tracker, *strawresponse, *kkbf_, kkmat_.strawMaterial(), pseedtraj, chcol, strawHitIdxs, strawhits, strawxings)){
-            // optionally (and if present) add the CaloCluster as a constraint
-            // verify the cluster looks physically reasonable before adding it TODO!  Or, let the KKCaloHit updater do it TODO
-            KKCALOHITCOL calohits;
-            if (kkfit_.useCalo() && hseed.caloCluster().isNonnull())kkfit_.makeCaloHit(hseed.caloCluster(),*calo_h, pseedtraj, calohits);
-            // set the seed range given the hits and xings
-            seedtraj.range() = kkfit_.range(strawhits,calohits,strawxings);
-            // create and fit the track
-            auto ktrk = make_unique<KKTRK>(config_,*kkbf_,seedtraj,fitpart,kkfit_.strawHitClusterer(),strawhits,strawxings,calohits);
-            // Check the fit
-            auto goodfit = goodFit(*ktrk);
-            // if we have an extension schedule, extend.
-            if(goodfit && exconfig_.schedule().size() > 0) {
-              kkfit_.extendTrack(exconfig_,*kkbf_, *tracker,*strawresponse, kkmat_.strawMaterial(), chcol, *calo_h, cc_H, *ktrk );
-              goodfit = goodFit(*ktrk);
-              // if finaling, apply that now.
-              if(goodfit && fconfig_.schedule().size() > 0){
-                ktrk->extend(fconfig_,nohits,noexings);
-                goodfit = goodFit(*ktrk);
+
+        // determine the fit direction hypothesis (or hypotheses)
+        auto helix_dirs = chooseHelixDir(hseed);
+        if(helix_dirs.empty()) continue; //bad helix, no fits to perform
+        const bool undefined_dir = helix_dirs.size() > 1; //fitting multiple hypotheses to determine the best fit
+        if(undefined_dir) ++nAmbiguous_;
+
+        // fit the track hypothesis (hypotheses)
+        std::unique_ptr<KKTRK> ktrk = nullptr;
+        float helix_dir(0.f); //selected track propagation direction
+        for(auto helix_dir_i : helix_dirs) {
+          const float dir = (helix_dir_i == HelixRecoDir::PropDir::downstream) ? 1.f : -1.f;
+          // Fit the track
+          auto ktrk_i = fitTrack(event, hseed,  dir, fpart_);
+          // check if a track is returned
+          if(ktrk_i) {
+            // if no track is selected yet, accept it, otherwise compare it to the current best fit
+            if(!ktrk) {
+              ktrk = std::move(ktrk_i);
+              helix_dir = dir;
+            } else {
+              const bool accept = compare_tracks(ktrk, ktrk_i) == 0;
```

Comment:
This logic would be clearer if you started with an array<std::unique_ptr<KKTRK>,2>, filled that in the loop over helix_dirs, and then called compare_tracks on the array.

Reply:
I updated to this style, passing the array to the compare tracks call.